src/_C004.cpp

```diff
@@ -1,136 +1,134 @@
-  } else {
-    for (uint8_t x = 0; x < element.valueCount; x++)
-    {
-      postDataStr += F("&field");
-      postDataStr += element.idx + x;
-      postDataStr += '=';
-      postDataStr += element.txt[x];
-    }
-  }
-  if (!ControllerSettings.UseDNS) {
-    // Patch the ControllerSettings to make sure we're using a hostname instead of an IP address
-    ControllerSettings.setHostname(F("api.thingspeak.com")); // PM_CZ: HTTP requests must contain host headers.
-    ControllerSettings.UseDNS = true;
-  }
-
-  int httpCode = -1;
-  send_via_http(
-    cpluginID,
-    ControllerSettings,
-    element._controller_idx,
-    F("/update"), // uri
-    F("POST"),
-    F("Content-Type: application/x-www-form-urlencoded\r\n"),
-    postDataStr,
-    httpCode);
-  return (httpCode >= 100) && (httpCode < 300);
-}
-
-#endif // ifdef USES_C004
+#include "src/Helpers/_CPlugin_Helper.h"
+#ifdef USES_C004
+
+// #######################################################################################################
+// ########################### Controller Plugin 004: ThingSpeak #########################################
+// #######################################################################################################
+
+# define CPLUGIN_004
+# define CPLUGIN_ID_004         4
+# define CPLUGIN_NAME_004       "ThingSpeak"
+
+bool CPlugin_004(CPlugin::Function function, struct EventStruct *event, String& string)
+{
+  bool success = false;
+
+  switch (function)
+  {
+    case CPlugin::Function::CPLUGIN_PROTOCOL_ADD:
+    {
+      ProtocolStruct& proto = getProtocolStruct(event->idx); //      = CPLUGIN_ID_004;
+      proto.usesMQTT     = false;
+      proto.usesAccount  = true;
+      proto.usesPassword = true;
+      proto.defaultPort  = 80;
+      proto.usesID       = true;
+      break;
+    }
+
+    case CPlugin::Function::CPLUGIN_GET_DEVICENAME:
+    {
+      string = F(CPLUGIN_NAME_004);
+      break;
+    }
+
+    case CPlugin::Function::CPLUGIN_INIT:
+    {
+      success = init_c004_delay_queue(event->ControllerIndex);
+      break;
+    }
+
+    case CPlugin::Function::CPLUGIN_EXIT:
+    {
+      exit_c004_delay_queue();
+      break;
+    }
+
+    case CPlugin::Function::CPLUGIN_GET_PROTOCOL_DISPLAY_NAME:
+    {
+      success = true;
+
+      switch (event->idx) {
+        case ControllerSettingsStruct::CONTROLLER_USER:
+          string = F("ThingHTTP Name");
+          break;
+        case ControllerSettingsStruct::CONTROLLER_PASS:
+          string = F("API Key");
+          break;
+        default:
+          success = false;
+          break;
+      }
+      break;
+    }
+
+    case CPlugin::Function::CPLUGIN_PROTOCOL_SEND:
+    {
+      if (C004_DelayHandler == nullptr) {
+        break;
+      }
+
+      if (C004_DelayHandler->queueFull(event->ControllerIndex)) {
+        break;
+      }
+
+      std::unique_ptr<C004_queue_element> element(new C004_queue_element(event));
```

Comment:
Why did you remove the `(std::nothrow)`?

Reply:
Hmm, tried to fix a large merge conflict, so I missed that. Will restore.